Settings with NVS backend

kapi-no opened the pull request:
Added support for NVS backend in Settings with following functionalities:

- full tree loading with settings_load()
- subtree loading with settings_load_subtree()
- subtree / single entry deletion with settings_delete()
- single entry update / write with settings_save_one()

Fixes #12160

kapi-no committed:
settings: adding new nvs backend

Added NVS backend to the Settings subsystem.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

nvlsianpu commented:
cc @Laczen

Laczen commented on subsys/settings/src/settings_init.c:
```diff
@@ -117,7 +117,55 @@ int settings_backend_init(void)
+	default_settings_nvs.cf_nvs.offset = DT_FLASH_AREA_STORAGE_OFFSET +
+		CONFIG_SETTINGS_NVS_OFFSET;
+
+	rc = nvs_init(&default_settings_nvs.cf_nvs, DT_FLASH_DEV_NAME);
```
I would move the nvs_init() into the settings_nvs_backend_init(), then a user can do the same procedure without having to use nvs as the default backend.

Laczen commented on subsys/settings/src/settings_init.c:
```diff
@@ -117,7 +117,55 @@ int settings_backend_init(void)
+	int rc;
+	struct flash_pages_info info;
+
+	rc = flash_get_page_info_by_offs(device_get_binding(DT_FLASH_DEV_NAME),
```
This can be removed, it is done in nvs_init()

Laczen commented on subsys/settings/Kconfig:
```diff
@@ -104,3 +110,25 @@ config SETTINGS_FS_MAX_LINES
+config SETTINGS_NVS_OFFSET
+	int "Offset of the NVS settings area"
+	default 0
+	depends on SETTINGS && SETTINGS_NVS
```
I would define SETTINGS_NVS_OFFSET to default to CONFIG_FLASH_STORAGE_OFFSET, but I don't know if this is possible.

Laczen commented on subsys/settings/Kconfig:
```diff
@@ -104,3 +110,25 @@ config SETTINGS_FS_MAX_LINES
+
+config SETTINGS_NVS_SECTOR_SIZE
+	int "Sector size of the NVS settings area"
+	default 4096
```
I would default this to DT_FLASH_ERASE_BLOCK_SIZE, if possible

carlescufi replied:
It should be possible by using `default $(dt_hex_val, DT_FLASH_ERASE_BLOCK_SIZE)`

Laczen replied:
Nice, I just learned something new and very interesting.

nvlsianpu commented on subsys/settings/Kconfig:
```diff
@@ -104,3 +110,25 @@ config SETTINGS_FS_MAX_LINES
 	help
 	  Limit how many items stored in a file before compressing
+
+config SETTINGS_NVS_SECTOR_SIZE
```
I would fetch sector size value from flash driver or flash_area API. Maybe keep just a multiplier of physical page size here (for support memories with a small pages by NVS.

nvlsianpu commented on subsys/settings/Kconfig:
```diff
@@ -104,3 +110,25 @@ config SETTINGS_FS_MAX_LINES
+	help
+	  The sector size to use for the NVS settings area in bytes.
+
+config SETTINGS_NVS_SECTOR_COUNT
```
as for fcb back-end this should be self corrected value. 
```
	  Number of areas to allocate in the settings FCB. A smaller number is
	  used if the flash hardware cannot support this value.
```

nvlsianpu replied:
Or just use whole storage area?

Laczen commented:
@kapi-no, I see a potential problem with the extension of the delete method to a subtree delete.

I like the feature, but if not all backends support this it will be required for the users to call delete for each item separately. If by accident a delete is called that does not match a entry but matches a subtree at the moment (present settings module) you would get an error that the delete failed. In the new method this delete will not return an error but all deletes called for correct entries will fail (because they have been deleted already).

So maybe it is better to keep the old functionality at the moment and to allow subtree delete only when all backends support it.

Laczen commented:
@nvlsianpu, I would keep all options (OFFSET, SECTORS, SECTOR_SIZE) to define the nvs backend and not drill them down to what is available in storage:

At the moment the configuration does two things: enable support for the backend **AND** setup a default backend, which makes it impossible to use two different backends at the same time. This is not good but I will not resubmit the changes I made for this.

Now by fixing everything to storage we are removing the possibility to use two backends of the same type, this is even worse.

It might seem simpler for the user, but IMO a better approach is to set good default values and to leave the ability to the user to use settings at it's full potential. As full potential I see:
a. A read-only backend (e.g. TLV based), for storing factory settings,
b. A rw backend like nvs, fcb, nffs for classical storage,
c. A rw backup backend like nvs, fcb, nffs for backing up in case something goes wrong with b.
...

nvlsianpu commented:
@kapi-no After rethinking - I would do this similar as for fcb-backend - associate storage with DT_FLASH_AREA_STORAGE_ID and use flash_area/map API to fetch all needed info. Keep nvs sector count and sector size configurable as well (NVS check sector size correctness on its own). I'm not sure whether sector size should be in absolute or relative unit.

@Laczen I cc'ed to you but didn't expect you response as you told that you don't want to involve in nvs back-end anymore. But good, this is open community. 

> Now by fixing everything to storage we are removing the possibility to use two back-ends of the same type, this is even worse.

I think the same: that should be possible to use multiple storage partitions with a few back-end, even of the same type. Unfortunately selection of storage is not working well now - (by Kconfig limitation). 
Regardless of that we want chose flash partition using a flash_map ID or using storage ID it doesn't work well.

We are considering functionalities a), b) c) you mentioned useful and want to introduce it in next iterations (so next patches).

Laczen commented:
> > Now by fixing everything to storage we are removing the possibility to use two back-ends of the same type, this is even worse.
> 
> I think the same: that should be possible to use multiple storage partitions with a few back-end, even of the same type. Unfortunately selection of storage is not working well now - (by Kconfig limitation).
> Regardless of that we want chose flash partition using a flash_map ID or using storage ID it doesn't work well.
>

If you just separate the default storage backend from the selected backends supported you can handle this problem limited to the default. Advanced users can then just disable the default backend and already do everything that is needed. Now the problem of the storage is pushed up by the selection of the backend. You could make the selection of a default backend select the required supported backend:
`CONFIG_SETTINGS_FCB=y` means `CONFIG_SETTINGS_FCB_DEFAULT=y` and `CONFIG_SETTINGS_FCB_SUPPORT=y` then nothing changes to the API, settings_init() can use `CONFIG_SETTINGS_FCB_DEFAULT` and cmake can use `CONFIG_SETTINGS_FCB_SUPPORT`.
`CONFIG_SETTINGS_CUSTOM=y` can be used to set according to specific needs.
 
> We are considering functionalities a), b) c) you mentioned useful and want to introduce it in next iterations (so next patches).

If you already know that this would be added, think how to do it now, and make sure that what is proposed now is inline with the future wishes (e.g. move nvs_init() to the nvs backend). Don't delay the decision because it will get more difficult as new backends are added.

kapi-no commented:
@nvlsianpu, @Laczen, I addressed most comments from your review.

Highlights:
- removed support for subtree deletion
- removed Kconfig configuration for the default instance of NVS, default configuration depends now on values fetched from flash_map API
- moved nvs_init(), so that custom backends can use their own configuration of NVS instance (instead of default one)

kapi-no committed:
tests: settings: adding nvs backend configuration to functional tests

Added the NVS backend configuration to functional tests of the Settings
module.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

lemrey commented:
In general, this looks okay but I worry that it will become costly with "many" `nvs` entries.
Do I recall correctly, that `nvs` reads all entries on lookups?

Laczen commented on subsys/settings/src/settings_nvs.c:
```diff
@@ -0,0 +1,238 @@
+	int rc;
+	u16_t last_name_id;
+
+	rc = nvs_init(&cf->cf_nvs, DT_FLASH_DEV_NAME);
```
The use of `DT_FLASH_DEV_NAME` limits the nvs file system to be on the main flash device, maybe it is better to pass this the `DT_FLASH_DEVICE` as a const char to `settings_nvs_backend_init()` or to extend settings_nvs with a `const char *flash_device_name`

kapi-no commented:
> In general, this looks okay but I worry that it will become costly with "many" nvs entries.
Do I recall correctly, that nvs reads all entries on lookups?

Yes, but the settings module usually loads all configurations at the start of the application. so it needs to go through all of them anyway.

However, using settings_load_subtree() to load specific configuration will be costly with many nvs entries, because - as you said - all entries will be read.

Laczen commented on subsys/settings/src/settings_init.c:
```diff
@@ -117,6 +117,52 @@ int settings_backend_init(void)
+	struct flash_sector sectors[NVS_MAX_SECTOR_NUM];
+
+	rc = flash_area_get_sectors(DT_FLASH_AREA_STORAGE_ID, &cnt, sectors);
+	if (rc) {
```
The default nvs system will take the complete storage area, so if a user would like to split up the storage area in two nvs systems (one for backup, one for runtime) this is not possible.

kapi-no replied:
It won't be an issue any more when DTS will let you define multiple storage areas for NVS.

Laczen replied:
True, but you will have to do changes the settings_init.c again to reflect this.

My BAG (big ambitious goal) would be to try and separate any knowledge about the dts from the settings module. I would do this by defining items in Kconfig and then filling them up in the configuration stage.

So how to do this: 
a. Define a settings_nvs_default_storage_label (default equal to storage)
b. Define a settings_nvs_default_sector_count (default equal to 65535, meaning it will take the maximum available sectors)
c. Define a settings_nvs_default_sector_size_multiplier (default to 1)
(d. optionally you could also define settings_nvs_default_dev_name)

Using these defines and flashmap you can find out the size of the storage area and adapt (using the size_multiplier) sector_count to reflect the real maximum. There would be no reference to the dts at all (except in Kconfig). This gives clear separation between dts and the settings module, which is good.

nvlsianpu replied:
^^ sounds good. I would fetch settings_nvs_default_dev_name using API (less to configure).
I prefer to fulfill this in next patch.

kapi-no replied:
I am going to add the multiplier to the Kconfig. It seems to be a really useful configuration option for devices with very small erase blocks. The rest can be updated in the future patch as @nvlsianpu said.

nvlsianpu replied:
^^OK

Laczen replied:
Maybe my proposal has been misunderstood, what I meant is define Kconfig variables:
`CONFIG_SETTINGS_NVS_DEFAULT_STORAGE_LABEL`, `CONFIG_SETTINGS_DEFAULT_SECTOR_COUNT` and `CONFIG_SETTINGS_DEFAULT_SECTOR_SIZE_MULTIPLIER` and use these to setup the default nvs backend. It is a very small change, I see no reason to delay this to a future patch.

kapi-no replied:
I see no value in defining CONFIG_SETTINGS_DEFAULT_SECTOR_COUNT. IMO, the number of sectors should cover the whole storage area that was assigned to the NVS backend. If the user wants to change the number of sectors within their application. they can override storage partition configuration using DTS overlays.

nvlsianpu replied:
It is impossible to transfer a label via Kconfig (I will welcome Kconfig enchantment for that). Kconfig is able to transfer strings as `CONFIG_<string-config> "a string"`.

Laczen replied:
> I see no value in defining CONFIG_SETTINGS_DEFAULT_SECTOR_COUNT. IMO, the number of sectors should cover the whole storage area that was assigned to the NVS backend. If the user wants to change the number of sectors within their application. they can override storage partition configuration using DTS overlays.

@kapi-no, the storage area was never assigned to the NVS backend, the NVS backend is taking it without any reasoning. I can understand that it would default to this but not to do it without any intervention possibility from the user. If `CONFIG_SETTINGS_NVS_DEFAULT_SECTOR_COUNT` would be defaulted to a large value it would result in the same user experience (default it is set to the complete area) but it leaves the user the possibility to override this decision.

@nvlsianpu, so it would be possible to set `CONFIG_SETTINGS_NVS_DEFAULT_LABEL = "STORAGE"`, this would be fine. Then you could use this to get the info from flash_map or not ?

nvlsianpu replied:
^^ No. Flash areas are identified by labeled numbers.

Laczen replied:
@nvlsianpu, this is bad. But hey, you know how I dislike flashmap, it does not provide any usability and is just taking up ram/flash.
To solve the problem we might be able to set the ID instead of the label, maybe using the trick from @carlescufi: `default $(dt_hex_val, DT_FLASH_AREA_STORAGE_ID)`

nvlsianpu replied:
^^ I know what you think about it.
> To solve the problem we might be able to set the ID instead of the label, maybe using the trick from @carlescufi: default $(dt_hex_val, DT_FLASH_AREA_STORAGE_ID)

Unfortunately this works only on project generation.

Laczen replied:
OK, once again I have given enough alternatives. If you make the decision that "it is the right of the settings module to take all the storage area" be brave enough to communicate this decision in a warning when the settings subsystem is selected. My proposal would be "We as the designers of the settings module have decided that the storage area is ours. This area cannot be used for anything else."

nvlsianpu replied:
@Laczen Why not creating another storage area instead of that. It will be dedicated to the application.

Now we have one partition:
```
		/* Storage partition will be used by FCB/NFFS/NVS if enabled. */
		storage_partition: partition@1000 {
			label = "storage";
			reg = <0x00001000 0x00010000>;
		};
```
This can be replaced on demand, if needs. Bellow NVS partition is truncated, and flowed by partition of user storage. Which looks legible. This solution doesn't mix Kconfig and DTS properties in order to get bounduaries for NVS - I think that's good.
```
		/* Storage partition will be used by FCB/NFFS/NVS if enabled. */
		storage_partition: partition@1000 {
			label = "storage";
			reg = <0x00001000 0x00008000>;
		};
		/* Storage partition will be used by user application. */
		custom_storage_partition: partition@9000 {
			label = "user-storage";
			reg = <0x00009000 0x00007000>;
		};
```

Laczen replied:
@nvlsianpu for two reasons:

1. DTS should be hardware only, at least as much as possible. If possible we should eliminate even the mcuboot, image0, image1, ... definitions. This is possible, using e.g. a Kconfig for flashmap. IMO it is not good to put even more software (logical volume) definitions in DTS
2. When changing the dts you need to do this in all board definitions -> large chance for errors. I know you can override these board definitions but you would still need to do that for all boards.

@nvlsianpu, I agree that you should not mix Kconfig and DTS properties to get/check the boundaries. By defining the label_id that will be used and using flash_map[label_id] for sizes, and flash_page_layout for page_size it is possible to set/check the boundaries without any DTS definition. To get the maximum sector count it is just incrementing a variable with the `sector_multiplier x page_size` until the size from flash_map[label_id] is reached.

nvlsianpu replied:
@Laczen 
comment to 1. I don't have strong preference on that where flash areas boundaries should be defined. Maybe worth to analyze how other environments deal with this (linux, Riot, MyNewt, mbed-os), sum-up this and work on that in dedicated issue. Lot of people will be interested in that. I feel that what you mention is a very good argument.
I don't know who have decided that flash areas are described via DTS - this idea might has its advocates. This will be <B>Switch to describe flash-partition via KConfig only</B> issue.
I'm keen to act on that if you want too. cc @hakonfam have you opinion on that.

comment to 2. Yes Kconfig is much less error prone. Honestly i was thinking about using override only when needed - not to changing any boards dts for that.

nvlsianpu replied:
@kapi-no Can you introduce CONFIG_SETTINGS_NVS_DEFAULT_SECTOR_COUNT back?

kapi-no replied:
yes

kapi-no replied:
done

nvlsianpu replied:
thanks

Laczen commented:
> In general, this looks okay but I worry that it will become costly with "many" `nvs` entries.
> Do I recall correctly, that `nvs` reads all entries on lookups?

Not exactly, nvs reads back until it finds a matching id (this is the last one). However, repeated lookup calls will make you read previously read entries again. This can make it worse than fcb where you now each entry will be read only once, but there will be repeated calls to the set routine.

I have been crushing my head on how to solve this problem, but I'm still crushing :-).

Laczen commented:
Oops, sorry for closing, it was not my intention.

nvlsianpu commented on subsys/settings/src/settings_init.c:
```diff
@@ -117,6 +117,53 @@ int settings_backend_init(void)
+	/* define the nvs file system using the page_info */
+	default_settings_nvs.cf_nvs.sector_size = sectors[0].fs_size;
+	default_settings_nvs.cf_nvs.sector_count = cnt;
+	default_settings_nvs.cf_nvs.offset = DT_FLASH_AREA_STORAGE_OFFSET;
```
```suggestion
	rc = flash_area_open(DT_FLASH_AREA_STORAGE_ID, &fa);
	if (rc) {
		zło;
	}
	default_settings_nvs.cf_nvs.offset = fa.fa_off;
	default_settings_nvs.flash_dev_name = flash_area_get_device(const struct flash_area *fa);
```

kapi-no replied:
It could be even simpler:

default_settings_nvs.flash_dev_name = fa.fa_dev_name

nvlsianpu replied:
default_settings_nvs.flash_dev_name = device_get_binding(fa.fa_dev_name);

nvlsianpu replied:
so actually exactly what is under the API.

nvlsianpu replied:
It's ok in the code.

lemrey commented on subsys/settings/src/settings_nvs.c:
```diff
@@ -0,0 +1,238 @@
+		if (name_id == NVS_NAMECNT_ID) {
+			break;
+		}
+		rc1 = nvs_read(&cf->cf_nvs, name_id, &name, sizeof(name));
```
I understand that saving the setting's name in one entry and the value in another (as opposed to saving everything in one entry) is done to optimize the usage of the flash, since once the application is compiled the name of the settings is basically static and it is only their value that changes. Thus, every time a setting's value is updated we will avoid rewriting the setting's name.

If my understanding is correct, it would be helpful to have a small comment here to remind why we are saving two entries instead of one.

kapi-no replied:
Yes, that's correct.

Added a short description to explain it.

lemrey reviewed (commented):
Looks good, left a few comments.

kapi-no commented on subsys/settings/src/settings_nvs.c:
```diff
@@ -0,0 +1,238 @@
+		}
+
+		if ((rc1 <= 0) || (rc2 <= 0)) {
+			/* NVS is dirty, clean it */
```
Added short description

kapi-no commented on subsys/settings/src/settings_nvs.c:
```diff
@@ -0,0 +1,238 @@
+	return 0;
+}
+
+/* ::csi_save implementation */
```
some kind of leftover, I removed it

kapi-no commented on subsys/settings/src/settings_nvs.c:
```diff
@@ -0,0 +1,238 @@
+	}
+
+	if (delete) {
+		return -ENOENT;
```
Yes, but I think it's better to distinguish between the entry deleted and entry not found cases.

kapi-no commented on subsys/settings/src/settings_nvs.c:
```diff
@@ -0,0 +1,238 @@
+			continue;
+		}
+
+		if ((delete) && (name_id == cf->last_name_id)) {
```
Yes.

This part of code only updates the NVS entry with NVS backend metadata. This entry contains the largest name ID that has ever been used.

Next if case does the actual reclaiming by deleting the setting item.

kapi-no commented on subsys/settings/include/settings/settings_nvs.h:
```diff
@@ -0,0 +1,49 @@
+extern "C" {
+#endif
+
+/* In nvs the settings data is stored as:
```
Rewrote it using your suggestions

lemrey commented on subsys/settings/src/settings_nvs.c:
```diff
@@ -0,0 +1,238 @@
+			continue;
+		}
+
+		if ((rc1 <= 0) || (rc2 <= 0)) {
```
Have you considered moving this check at initialization time, so that when loading a setting we can skip the second read (for the setting's value) if the setting's name is not found?

kapi-no replied:
> Have you considered moving this check at initialization time, so that when loading a setting we can skip the second read (for the setting's value) if the setting's name is not found?

No, I haven't, but it would make settings_load a bit faster. However, load function is usually called only once, so I wonder if this is worth doing.

> Generally, when could that situation happen?

Well, this is not so common, but it could happen when the device is reset during a write operation of settings value.

@nvlsianpu, @Laczen, any thoughts about this?
I would leave this as it is now (safer & slower). We can always make optimizations afterwards if somebody starts complaining about NVS performance.

lemrey replied:
I am fine with addressing with afterwards if necessary.

nvlsianpu commented:
@Laczen Can you accept the PR?

@jhedberg @Vudentz @fnde-ot I want to ask to merge this PR. Have anyone more comments?

Laczen commented:
@nvlsianpu, @kapi-no, I am still not convinced about the backend-init method for NVS. It is still using a mix of dts/flashmap. What exactly does flash_area_get_sectors() return when cnt is set to 1 (documentation about the meaning of sectors is limited) ? It seems like the code is settings cnt based upon known variables (page size) and is returning this page size again.

Laczen commented:
@nvlsianpu, @kapi-no, my proposal would be:

```
int settings_backend_init(void)
{
	int rc;
	u16 cnt = 0;
	size_t nvs_sector_size, nvs_size = 0;
	const struct flash_area *fa;

	rc = flash_area_open(DT_FLASH_AREA_STORAGE_ID, &fa);
	if (rc) {
		return rc;
	}
	/* DT_FLASH_STORAGE_BLOCK_SIZE could be replaced by flash_page request */
	nvs_sector_size = CONFIG_SETTINGS_NVS_SECTOR_SIZE_MULT * 	
			  DT_FLASH_AREA_STORAGE_BLOCK_SIZE;
	while ((cnt < CONFIG_SETTINGS_NVS_SECTOR_COUNT) && (nvs_size < fa->size)) {
		nvs_size += nvs_sector_size
		cnt++;
	}

	/* define the nvs file system using the page_info */
	default_settings_nvs.cf_nvs.sector_size = nvs_sector_size;
	default_settings_nvs.cf_nvs.sector_count = cnt;
	default_settings_nvs.cf_nvs.offset = fa->fa_off;
	default_settings_nvs.flash_dev_name = fa->fa_dev_name;

	rc = settings_nvs_backend_init(&default_settings_nvs);
	if (rc) {
		return rc;
	}

	rc = settings_nvs_src(&default_settings_nvs);

	if (rc) {
		return rc;
	}

	rc = settings_nvs_dst(&default_settings_nvs);

	return rc;
}
```

kapi-no commented:
> @nvlsianpu, @kapi-no, I am still not convinced about the backend-init method for NVS. It is still using a mix of dts/flashmap. What exactly does flash_area_get_sectors() return when cnt is set to 1 (documentation about the meaning of sectors is limited) ? It seems like the code is setting cnt based upon known variables (page size) and is returning this page size again.

Yes, it returns the number of available pages. I am sure that there are many ways to implement this, but I prefer using the Flash Map API to count the number of pages instead of doing it on my own.

Laczen commented:
> > @nvlsianpu, @kapi-no, I am still not convinced about the backend-init method for NVS. It is still using a mix of dts/flashmap. What exactly does flash_area_get_sectors() return when cnt is set to 1 (documentation about the meaning of sectors is limited) ? It seems like the code is setting cnt based upon known variables (page size) and is returning this page size again.
> 
> Yes, it returns the number of available pages. I am sure that there are many ways to implement this, but I prefer using the Flash Map API to count the number of pages instead of doing it on my own.

@kapi-no, there is no counting of the number of pages that are inside, there is a counting of the number of nvs_sectors. I prefer counting over a divide.

If flash_area_get_sectors(.., cnt=1, ...) returns the page size you can replace DT_FLASH_STORAGE_ERASE_BLOCK_SIZE with a call to flash_area_get_sectors(), I don't mind. But test it first because it might return the complete storage area. If you first need to set cnt based upon ERASE_BLOCK_SIZE to get the ERASE_BLOCK_SIZE then don't do this, it is just waste.

ioannisg commented on subsys/settings/Kconfig:
```diff
@@ -104,3 +110,18 @@ config SETTINGS_FS_MAX_LINES
+	default 8
+	depends on SETTINGS && SETTINGS_NVS
+	help
+	  Number of sectors used for the NVS settings area
```
Something wrong with new line, perhaps?
@kapi-no

kapi-no replied:
Do you want me to add a blank line at the end of this file?

ioannisg replied:
Yes, I think we should have blank lines at the end of all source files

ioannisg replied:
So we don't get these GH error signs.

kapi-no replied:
okay, done
